azurerm/resource_arm_operational_insight_workspace.go

```diff
@@ -0,0 +1,224 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+
+	"github.com/Azure/azure-sdk-for-go/arm/operationalinsights"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceArmOperationalInsightWorkspaceService() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmOperationalInsightWorkspaceCreateUpdate,
+		Read:   resourceArmOperationalInsightWorkspaceRead,
+		Update: resourceArmOperationalInsightWorkspaceCreateUpdate,
+		Delete: resourceArmOperationalInsightWorkspaceDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRmOperationalInsightWorkspaceName,
+			},
+			"location": {
+				Type:      schema.TypeString,
+				Required:  true,
+				ForceNew:  true,
+				StateFunc: azureRMNormalizeLocation,
+			},
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"workspace_id": { // a.k.a. customer_id
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"portal_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sku": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"retention_in_days": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+			},
+			"primary_shared_key": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"secondary_shared_key": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmOperationalInsightWorkspaceCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).workspacesClient
+	log.Printf("[INFO] preparing arguments for AzureRM Operational Insight workspace creation.")
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+
+	skuName := d.Get("sku")
+	sku, err := getSku(skuName)
+	if err != nil {
+		return err
+	}
+
+	retentionInDays := int32(d.Get("retention_in_days").(int))
+
+	tags := d.Get("tags").(map[string]interface{})
+
+	parameters := operationalinsights.Workspace{
+		Name:     &name,
+		Location: &location,
+		Tags:     expandTags(tags),
+		WorkspaceProperties: &operationalinsights.WorkspaceProperties{
+			Sku:             sku,
+			RetentionInDays: &retentionInDays,
+		},
+	}
+
+	cancel := make(chan struct{})
+	workspaceChannel, error := client.CreateOrUpdate(resGroup, name, parameters, cancel)
```

Comment:
two things:

1) can we in-line the cancel channel, since we're not making use of it at this time
2) we'd normally ignore the response from the Create (in this case `workspaceChannel`) and instead do a Get request to get the object and it's ID (since the URI's _could_ be different (as is commented out) - is there any reason why we've taken this route instead?

Reply:
I've done.  

I'm new to go-lang. However, I don't like copy-and-paste programming without understanding the reason. Since I just couldn't understand why you don't use the return value of CreateOrUpdate.  Could you tell me the reason for my learning? :)